apps/events/models.py

```diff
@@ -0,0 +1,109 @@
+from datetime import datetime, timedelta
+
+from django.db import models
+from pytz import UTC
+
+from apps.chat.models import ChatMessageType
+from apps.events.actions import log
+from apps.experiments.models import Experiment, ExperimentSession, SessionStatus
+from apps.utils.models import BaseModel
+
+ACTION_FUNCTIONS = {"log": log}
+
+
+class EventActionType(models.TextChoices):
+    LOG = "log"  # Prints the last message
+    SUMMARIZE = "summarize"  # requires prompt
+
+
+class EventAction(BaseModel):
+    action_type = models.CharField(choices=EventActionType.choices)
+    params = models.JSONField(null=True, blank=True)  # The parameters for the specific action
+
+
+class BaseTrigger(BaseModel):
+    action = models.OneToOneField(EventAction, on_delete=models.CASCADE, related_name="action")
+
+    class Meta:
+        abstract = True
+
+    def fire(self, session):
+        return ACTION_FUNCTIONS[self.action.action_type](session)
+
+
+class TimeoutTrigger(BaseTrigger):
+    experiment = models.ForeignKey(Experiment, on_delete=models.CASCADE, related_name="timeout_triggers")
+    delay = models.PositiveIntegerField(
+        help_text="The amount of time in seconds to fire this trigger.",
+    )
+    total_num_triggers = models.IntegerField(
+        default=1,
+        help_text="The number of times to fire this trigger",
+    )
+    end_conversation = models.BooleanField(
+        default=True,
+        help_text="Whether to end the conversation after all the triggers have expired",
+    )
+
+    def timed_out_sessions(self):
+        sessions = []
+        trigger_time = datetime.now().astimezone(UTC) - timedelta(seconds=self.delay)
+        open_experiment_sessions = self.experiment.sessions.filter(ended_at=None)
+        for session in open_experiment_sessions:
+            try:
+                if self.stats.get(session=session).trigger_count >= self.total_num_triggers:
+                    continue
+            except TriggerStats.DoesNotExist:
+                # This trigger didn't run yet
+                pass
+            last_human_message = session.chat.messages.filter(message_type=ChatMessageType.HUMAN).last()
+            if last_human_message and last_human_message.created_at < trigger_time:
```

Comment:
Could all this logic be done in the DB? Maybe similar to https://github.com/dimagi/open-chat-studio/pull/288?